paasta_tools/cleanup_kubernetes_cr.py

```diff
@@ -0,0 +1,142 @@
+# limitations under the License.
+"""
+Usage: ./cleanup_kubernetes_cr.py [options]
+
+Command line options:
+
+- -d <SOA_DIR>, --soa-dir <SOA_DIR>: Specify a SOA config dir to read from
+- -v, --verbose: Verbose output
+"""
+import argparse
+import logging
+import sys
+from typing import Sequence
+
+from paasta_tools.kubernetes_tools import CustomResource
+from paasta_tools.kubernetes_tools import delete_custom_resource
+from paasta_tools.kubernetes_tools import KubeClient
+from paasta_tools.kubernetes_tools import list_custom_resources
+from paasta_tools.kubernetes_tools import load_custom_resources
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import load_all_configs
+from paasta_tools.utils import load_system_paasta_config
+
+log = logging.getLogger(__name__)
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description='Cleanup custom_resources.')
+    parser.add_argument(
+        '-d', '--soa-dir', dest="soa_dir", metavar="SOA_DIR",
+        default=DEFAULT_SOA_DIR,
+        help="define a different soa config directory",
+    )
+    parser.add_argument(
+        '-v', '--verbose', action='store_true',
+        dest="verbose", default=False,
+    )
+    parser.add_argument(
+        '-c', '--cluster', default=None,
+        help="Cluster to cleanup CRs for",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def main() -> None:
+    args = parse_args()
+    soa_dir = args.soa_dir
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+    else:
+        logging.basicConfig(level=logging.INFO)
+
+    kube_client = KubeClient()
+
+    system_paasta_config = load_system_paasta_config()
+    cluster = args.cluster or system_paasta_config.get_cluster()
+    custom_resources = load_custom_resources(system_paasta_config)
+    cleanup_kube_succeeded = cleanup_all_custom_resources(
+        kube_client=kube_client,
+        soa_dir=soa_dir,
+        cluster=cluster,
+        custom_resources=custom_resources,
+    )
+    sys.exit(0 if cleanup_kube_succeeded else 1)
+
+
+def cleanup_all_custom_resources(
+    kube_client: KubeClient,
+    soa_dir: str,
+    cluster: str,
+    custom_resources: Sequence[CustomResource],
+) -> bool:
+    cluster_crds = {
+        crd.spec.names.kind
+        for crd in
+        kube_client.apiextensions.list_custom_resource_definition(
+            label_selector="yelp.com/paasta_service",
+        ).items
+    }
+    log.debug(f"CRDs found: {cluster_crds}")
+    results = []
+    for custom_resource in custom_resources:
+        if custom_resource.kube_kind.singular not in cluster_crds:
+            # TODO: kube_kind.singular seems to correspond to `crd.names.kind`
+            # and not `crd.names.singular`
+            log.warning(
+                f"CRD {custom_resource.kube_kind.singular} "
+                f"not found in {cluster}",
+            )
+            continue
+        config_dicts = load_all_configs(
```

Comment:
i'm surprised that we've added another way of doing this.

can we just use the existing `get_service_instance_list` from `paasta_tools.utils`? I think it's slightly different in that it only returns the instance list, rather than the body of the yaml, but I think that's ok - we only seem to use that anyway?

I think you'll have to add `CassandraCluster` to `paasta_tools.utils.INSTANCE_TYPES` to use it though

Reply:
I think this was a deliberate choice so that it wasn't coupled to paasta knowing about all the different instance types...

I'm not sure what's better to be honest. But imagine you have an operator and don't want anything to do with paasta but just want something to sync CRs.

Reply:
but this thing only queries custom resources with the `yelp.com/paasta_service` label as a qualifier, so presumably they're being looked after by paasta anyway in that case?

yeah, it's kind of odd to me in a place where we're only doing half the job - I can't imagine that case.

Reply:
Can we do that as a part of another CR please?
This one is already becomes too big for a just additional of simple script.

Reply:
yeah I guess so